feat: Use quota in volume size option

dmga44 opened the pull request:
The following changes are proposed:
- introduce the option ```usequotainvolumesize``` to display the user specific quota instead of the storage total size. This option is also added to tweaks hidden file.
- introduce the option ```statfscachetimeout``` to tune the statfs output caching, it is disabled by default. This option is also added to tweaks hidden file.
- add a test to check the behavior of the option.
- a small refactor in sauna_client.h (a new option is added and it would be nice to pack it together).

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
feat: Use quota in volume size option

This commit adds a new option to the client which allows users to see
as the total space of the drive the assigned quota to themselves (user
and group quota) if enabled. This option is also added to the tweaks
hidden file.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
test: Test use quota in volume size option

This commit adds the test_use_quota_in_volume_size to the
ShortSystemTests suite.

Signed-off-by: Dave <dave@leil.io>

dmga44 committed:
refactor: Consolidate platform specific code

The following changes are proposed:
- enclose limit_glibc_malloc_arenas option and its default value in
Linux specific parts of the code.
- reduce the number of ifdef/endif processor directives by
consolidating the platform specific lines of code.

The goal of this commit is to increase readability of this part of the
code.

Signed-off-by: Dave <dave@leil.io>

lgsilva3087 commented on tests/test_suites/ShortSystemTests/test_use_quota_in_volume_size.sh:
```diff
@@ -0,0 +1,35 @@
+# Test will only succeed if the available space is more than 1GiB
+#
+# To run this test you need to add the following lines to /etc/sudoers.d/saunafstest:
+#
+# saunafstest ALL = NOPASSWD: /usr/bin/tee /tmp/SaunaFS-autotests/mnt/sfs*
+#
+#
```
Consider adding this to the setup machine script.

dmga44 replied:
This was copied from another test, so it should be there (the ```tee``` line in sudoers.d).

lgsilva3087 replied:
In this case, we can remove the comments from the test.

dmga44 replied:
Removed.

lgsilva3087 commented on src/mount/fuse/mount_config.cc:
```diff
@@ -239,6 +240,8 @@ void usage(const char *progname) {
 "    -o limitglibcmallocarenas=N  limit glibc malloc arenas to given value - prevents "
 				"from using huge amount of virtual memory. Use it in constrained memory "
 				"environments (default: %u)\n"
+"    -o usequotainvolumesize=0|1  replace normal volume size with the space hard limit "
+				"of the user specific quota (if assigned) when set to 1 (default: %u)\n"
```
Consider:
when set to 1, use the user-specific quota hard limit as the volume size.

dmga44 replied:
Applied.

lgsilva3087 commented on src/mount/fuse/mount_config.cc:
```diff
@@ -239,6 +240,8 @@ void usage(const char *progname) {
 "    -o limitglibcmallocarenas=N  limit glibc malloc arenas to given value - prevents "
 				"from using huge amount of virtual memory. Use it in constrained memory "
 				"environments (default: %u)\n"
+"    -o usequotainvolumesize=0|1  replace normal volume size with the space hard limit "
+				"of the user specific quota (if assigned) when set to 1 (default: %u)\n"
```
Does it consider quota for groups?

dmga44 replied:
I considered it, even coded the idea but the context coming from fuse layer passes a wrong ```gid```. Have no idea why.

lgsilva3087 replied:
Please check about secondary groups. Specifically this constant kSecondaryGroupsBit.

dmga44 replied:
Thanks.

I checked that only the main group of the user was the one affected, so I only asked for the main group.

lgsilva3087 commented on src/mount/sauna_client.cc:
```diff
@@ -687,6 +688,34 @@ struct statvfs statfs(Context &ctx, Inode ino) {
 	}
 	(void)ino;
 	fs_statfs(&totalspace,&availspace,&trashspace,&reservedspace,&inodes);
+	if (gUseQuotaInVolumeSize) {
```
Applications can trigger a surprisingly high number of statfs (even bigger in Windows). Consider to cache the final quota value for a short time (1-5 seconds).

dmga44 replied:
Well, I gave it some thought and the idea could be extended to the whole statfs output because the quota result (specially user's used space) could change at the same pace of the whole storage used space.

I have the cache for the whole statfs and an option to tune it, and also added it to the tweaks, so it could be activated in runtime.

lgsilva3087 replied:
Great.

lgsilva3087 reviewed (approved):
Looks good, but please see my comments.

rolysr commented on tests/test_suites/ShortSystemTests/test_use_quota_in_volume_size.sh:
```diff
@@ -0,0 +1,35 @@
+cd "${info[mount0]}"
+hard_limit=$((1024*1024*1024))
+initial_size=$(df --block-size=1 | grep "${info[mount0]}" | awk '{print $2}')
+[[ ${initial_size} -ne ${hard_limit} ]] && echo "OK: the available space is different than 1GiB"
```
As this test is supposed to pass when available space is greater than 1GiB, then you can consider using `assert_greater_than` for better readability

dmga44 replied:
Applied.

rolysr reviewed (approved):
LGTM

dmga44 committed:
feat: Statfs cache

The statfs syscall output could be cached for each user requesting it.
The main goal is to prevent a flooding of unnecessary calls in certain
workloads.

The commit includes the option to tune it and adding it to the tweaks
hidden file parameters.

Signed-off-by: Dave <dave@leil.io>
